CMakeLists.txt

```diff
@@ -198,125 +206,208 @@ else()
+        ${PTLS_INCLUDE_DIRS}
+        ${OPENSSL_INCLUDE_DIR}
+    PUBLIC
+        picoquic)
 
 target_link_libraries(picoquic-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-)
+    PRIVATE
+        ${OPENSSL_LIBRARIES}
+    PUBLIC
+        ${PTLS_LIBRARIES}
+        Threads::Threads)
 
-add_library(picoquic-log
-    ${LOGLIB_LIBRARY_FILES}
-)
+set_picoquic_build_options(picoquic-core)
 
-target_include_directories(picoquic-log PUBLIC loglib)
+add_library(picoquic-log ${LOGLIB_LIBRARY_FILES})
 
-add_library(picohttp-core
-    ${PICOHTTP_LIBRARY_FILES}
-)
+target_include_directories(picoquic-log
+    PRIVATE
+        ${PTLS_INCLUDE_DIRS}
+    PUBLIC
+        picoquic
+        loglib)
+
+add_library(picohttp-core ${PICOHTTP_LIBRARY_FILES})
 
 target_link_libraries(picohttp-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-)
-
-add_executable(picoquicdemo
-    picoquicfirst/picoquicdemo.c
-    picoquicfirst/getopt.c
-)
-
-target_link_libraries(picoquicdemo
-    picoquic-log
-    picoquic-core
-    picohttp-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picoquicdemo PRIVATE loglib picohttp)
-
-add_executable(picolog_t
-    picolog/picolog.c
-)
-
-target_link_libraries(picolog_t
-    picoquic-log
-    picoquic-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picolog_t PRIVATE loglib)
-
-add_executable(picoquic_ct picoquic_t/picoquic_t.c
-    ${PICOQUIC_TEST_LIBRARY_FILES}
-)
-
-target_link_libraries(picoquic_ct
-    picoquic-log
-    picoquic-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picoquic_ct PRIVATE loglib)
-
-add_executable(picohttp_ct picohttp_t/picohttp_t.c
-    ${PICOQUIC_TEST_LIBRARY_FILES}
-    ${PICOHTTP_TEST_LIBRARY_FILES}
-)
-
-target_link_libraries(picohttp_ct
-    picoquic-log
-    picoquic-core
-    picohttp-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picohttp_ct PRIVATE loglib picohttp)
-
-add_executable(picoquic_sample
-    sample/sample.c
-    sample/sample_client.c
-    sample/sample_server.c
-)
-
-target_link_libraries(picoquic_sample
-    picoquic-log
-    picoquic-core
-    ${PTLS_LIBRARIES}
-    ${OPENSSL_LIBRARIES}
-    ${CMAKE_DL_LIBS}
-    ${CMAKE_THREAD_LIBS_INIT}
-)
-
-target_include_directories(picoquic_sample PRIVATE loglib picoquic)
-
-set(TEST_EXES picoquic_ct picohttp_ct)
+    PRIVATE
+        ${PTLS_LIBRARIES}
+        ${OPENSSL_LIBRARIES}
+    PUBLIC
+        picoquic-core)
+
+target_include_directories(picohttp-core
+    PRIVATE
+        ${PTLS_INCLUDE_DIRS}
+        ${OPENSSL_INCLUDE_DIR}
+    PUBLIC
+        picoquic)
+
+set_picoquic_build_options(picohttp-core)
+
+include(CTest)
+
+if(BUILD_TESTING AND picoquic_BUILD_TESTS)
+
```

Comment:
I wonder whether we need to do this kind of selection using CMake options, or whether we should instead use CMake to define a series of individual "make" targets, such as defining a "core" target with the essential libraries, a "test" target, a "demo", a "sample" target, and maybe an "all" target for compatibility with the current behavior.

When building through fetch, don't we have the opportunity to do something like this?
```
cmake build
cd build
make core
```

Reply:
It is possible to do "make picoquic-core".  It will build it and any dependencies.